internal/util/kubernetes/unstructured_util.go

```diff
@@ -1,13 +1,309 @@
+	restConfig *rest.Config,
+	object *GenericObject,
+	pluralName string,
+) error {
+
+	unstruc, err := ObjectToUnstructured(object)
+	if err != nil {
+		return err
+	}
+
+	gvr, err := getGroupVersionResource(object, pluralName)
+	if err != nil {
+		return err
+	}
+
+	return UpdateUnstructuredCR(ctx, restConfig, unstruc, gvr, object.Namespace, object.Name)
+}
+
+func UpdateUnstructuredCR(
+	ctx context.Context,
+	restConfig *rest.Config,
+	unstruc *unstructured.Unstructured,
+	gvr schema.GroupVersionResource,
+	namespace string,
+	name string,
+) error {
+	numaLogger := logger.FromContext(ctx)
+	numaLogger.Debugf("will update resource %s/%s of type %+v", namespace, name, gvr)
+
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create dynamic client: %v", err)
+	}
+
+	_, err = client.Resource(gvr).Namespace(namespace).Update(ctx, unstruc, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update resource %s/%s of type %+v, err=%v", namespace, name, gvr, err)
+	}
+
+	numaLogger.Infof("successfully updated resource %s/%s of type %+v", namespace, name, gvr)
+	return nil
+}
+
+func parseApiVersion(apiVersion string) (string, string, error) {
+	// should be separated by slash
+	index := strings.Index(apiVersion, "/")
+	if index == -1 {
+		// if there's no slash, it's just the version, and the group should be "core"
+		return "core", apiVersion, nil
+	} else if index == len(apiVersion)-1 {
+		return "", "", fmt.Errorf("apiVersion incorrectly formatted: unexpected slash at end: %q", apiVersion)
+	}
+	return apiVersion[0:index], apiVersion[index+1:], nil
+}
+
+func ObjectToUnstructured(object *GenericObject) (*unstructured.Unstructured, error) {
+	var asMap map[string]any
+	err := util.StructToStruct(object, &asMap)
+	if err != nil {
+		return nil, err
+	}
+
+	return &unstructured.Unstructured{Object: asMap}, nil
+}
+
+func UnstructuredToObject(u *unstructured.Unstructured) (*GenericObject, error) {
+	var genericObject GenericObject
+	err := util.StructToStruct(u, &genericObject)
+	if err != nil {
+		return nil, err
+	}
+
+	return &genericObject, err
+}
+
+func UpdateStatus(ctx context.Context, restConfig *rest.Config, object *GenericObject, pluralName string) error {
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create dynamic client: %v", err)
+	}
+
+	gvr, err := getGroupVersionResource(object, pluralName)
+	if err != nil {
+		return err
+	}
+
+	// Rationale for commenting this out:
+	// If we get a conflict, it means that the Status we were starting from in this reconciliation was old.
+	// If the new Status is derived from the old Status, then that would've been an invalid state.
+	// todo: since the primary motivation for this retry logic was to avoid errors showing up in the log, we may be
+	// able to instead conditionally log a warning vs an error for that
+	//err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
@afugazzotto I wanted to bring it to your attention that I commented this out. Any time our `Reconcile()` function returns an error, there will be an automatic retry. 

I think in general with "retry on conflict", the solution is to Get the current object and then do all of the logic that you need to do to correctly update that object from its current state. My concern is this type of scenario:
1. In the previous reconciliation, one of our `Status.Conditions` was updated.
2. Our informer for the `Status` is slow, so for our next reconciliation we have a somewhat outdated version of the `Status` which doesn't include that previous update.
3. Our `object.Status` below doesn't include that update, but then it gets set in the object

Reply:
I introduced the `RetryOnConflict` with [this PR](https://github.com/numaproj/numaplane/pull/62) to fix the error we were seeing when updating state: `Operation cannot be fulfilled on ...: the object has been modified; please apply your changes to the latest version and try again`. If with your queueing implementation the error is no longer present even without the `RetryOnConflict` logic, it can be removed.

Reply:
yeah, I recall it was because we were seeing that error in the log, and even though it would retry, it would be alarming to users to see an error. Let me look into addressing that "false alarm" error separately...